### PR TITLE
Add missing method to get all datasources

### PR DESCRIPTION
**Pull request by akamensky**

### CODE REVIEW

File: datasource.go [hunk 1]
```diff
@@ -150,3 +150,14 @@ func (c *Client) DeleteDataSource(id int64) error {
...
+
+// DataSources returns all data sources as defined in Grafana
+func (c *Client) DataSources() ([]*DataSource, error) {
+	result := make([]*DataSource, 0)
+	err := c.request("GET", "/api/datasources", nil, nil, &result)
+	if err != nil {
+		return nil, err
+	}
+
+	return result, err
```
Comment (aknuds1) [inline datasource.go]:
```suggestion
	return result, nil
```

File: datasource.go [hunk 2]
```diff
@@ -150,3 +150,14 @@ func (c *Client) DeleteDataSource(id int64) error {
 
 	return c.request("DELETE", path, nil, nil, nil)
 }
+
+// DataSources returns all data sources as defined in Grafana
```
Comment (aknuds1) [inline datasource.go]:
```suggestion
// DataSources returns all data sources as defined in Grafana.
```

**Review by aknuds1 — DISMISSED** [review]
Have to check if this is something we want, but saw some things that should be fixed.

**Comment by akamensky** [comment]
FWIW other than period at the end of the sentence the rest was just copy/paste from code right above it...

**Comment by akamensky** [comment]
> Have to check if this is something we want, but saw some things that should be fixed.

I mean, you have methods to create/delete/update datasources, but no method to get their list??? I am using my own fork vendored in, just because your CRUD interface is not complete.

**Comment by aknuds1** [comment]
> FWIW other than period at the end of the sentence the rest was just copy/paste from code right above it...

That doesn't mean we shouldn't make improvements.

**Comment by akamensky** [comment]
@aknuds1 please review the changes as your requested.

**Comment by aknuds1** [comment]
I already said we need to discuss internally before deciding:

> Have to check if this is something we want

Please have some patience.

**Comment by akamensky** [comment]
> I already said we need to discuss internally before deciding:
> 
> Please have some patience.

I am asking to review the changes you yourself requested, not the merging and of the changes. Do take your time with discussion if that's the policy. And no need to be rude about it.

**Comment by aknuds1** [comment]
I'm not being rude, far from it. We will get back to you when we've had the time to review the feature addition.

**Comment by jinjaghost** [comment]
@akamensky: what do you think of adding the custom header "x-grafana-org-id" to make this method capable of listing datasources for a given OrgId ? This is the way the Grafana SPA uses the datasources API call to request list for a given org.

**Comment by akamensky** [comment]
@jinjaghost that would be inconsistent with other queries in this library. And also that should be it's own method in library as it won't be fetching all data sources, but their subset. Something named like "GetDataSourcesByOrgId".

**Review by aknuds1 — COMMENTED** [review]
We've discussed it and decided that even though the client is in deprecated mode now, we can accept the PR. Are you able to add a test though?

**Comment by Etherdaemon** [comment]
> We've discussed it and decided that even though the client is in deprecated mode now, we can accept the PR. Are you able to add a test though?

Hi @aknuds1 - when you say this client is in deprecated mode, what is the alternative client to be used with terraform-provider-grafana? I was planning on putting a bit of work into adding the new alert v2 parts from Grafana v8.3.X

**Comment by davidnewhall** [comment]
This doesn't really seem deprecated?

**Comment by jalevin** [comment]
I believe this is now implemented. https://github.com/grafana/grafana-api-golang-client/blob/9c38b6f437bb9688f655ace9e66cc8b8f91ce0f4/datasource.go#L371

**Pull request closed without merging**